test/Microsoft.AspNetCore.Server.KestrelTests/ChunkedRequestTests.cs

```diff
@@ -0,0 +1,477 @@
+                }
+
+                response.Headers.Clear();
+                response.Headers["Content-Length"] = new[] { "11" };
+
+                await response.Body.WriteAsync(Encoding.ASCII.GetBytes("Hello World"), 0, 11);
+            }, testContext))
+            {
+                using (var connection = new TestConnection(server.Port))
+                {
+                    await connection.Send(
+                    "POST / HTTP/1.1",
+                    "Transfer-Encoding: chunked",
+                    "",
+                    "Cio",
+                    "HelloChunked",
+                    "0",
+                    "");
+
+                    // Should really be a 40x as is bad request
+                    await connection.Receive(
+                        "HTTP/1.1 500 Internal Server Error",
+                        "");
+                    await connection.ReceiveStartsWith("Date:");
+                    await connection.ReceiveEnd(
+                        "Content-Length: 0",
+                        "Server: Kestrel",
+                        "",
+                        "");
+                }
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(ConnectionFilterData))]
+        public async Task InvalidSizedDataResultsIn500(ServiceContext testContext)
+        {
+            using (var server = new TestServer(async httpContext =>
+            {
+                var response = httpContext.Response;
+                var request = httpContext.Request;
+
+                var buffer = new byte[200];
+
+                while (await request.Body.ReadAsync(buffer, 0, buffer.Length) != 0)
+                {
+                    ;// read to end
+                }
+
+                response.Headers.Clear();
+                response.Headers["Content-Length"] = new[] { "11" };
+
+                await response.Body.WriteAsync(Encoding.ASCII.GetBytes("Hello World"), 0, 11);
+            }, testContext))
+            {
+                using (var connection = new TestConnection(server.Port))
+                {
+                    await connection.Send(
+                        "POST / HTTP/1.1",
+                        "Transfer-Encoding: chunked",
+                        "",
+                        "C",
+                        "HelloChunkedInvalid",
+                        "0",
+                        "");
+
+                    // Should really be a 40x as is bad request
+                    await connection.Receive(
+                        "HTTP/1.1 500 Internal Server Error",
+                        "");
+                    await connection.ReceiveStartsWith("Date:");
+                    await connection.ReceiveEnd(
+                        "Content-Length: 0",
+                        "Server: Kestrel",
+                        "",
+                        "");
+                }
+            }
+        }
+
+        private class TestApplicationErrorLogger : ILogger
```

Comment:
Can we move these private classes to TestHelpers? As simple as these are, I don't like having exact copies that could diverge.